Editor/Scripts/MeshSyncServerInspector.cs

```diff
@@ -1,12 +1,81 @@
-using Unity.FilmInternalUtilities;
+using System.Collections.Generic;
 using Unity.FilmInternalUtilities.Editor;
 using UnityEditor;
 using UnityEngine;
 
 namespace Unity.MeshSync.Editor  {
 [CustomEditor(typeof(MeshSyncServer))]
-internal class MeshSyncServerInspector : BaseMeshSyncInspector   {
-    
+[InitializeOnLoad]
+internal class MeshSyncServerInspector : BaseMeshSyncInspector {
+#if AT_USE_PROBUILDER
+    static System.Reflection.FieldInfo versionInfo = typeof(UnityEngine.ProBuilder.ProBuilderMesh).GetField("m_VersionIndex", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+
+    static Dictionary<UnityEngine.ProBuilder.ProBuilderMesh, ushort> versionIndices = new Dictionary<UnityEngine.ProBuilder.ProBuilderMesh, ushort>();
+
+    static MeshSyncServerInspector()
+    {
+        BaseMeshSync.ProBuilderBeforeRebuild += ProBuilderBeforeRebuild;
+        BaseMeshSync.ProBuilderAfterRebuild += ProBuilderAfterRebuild;
+    }
+
+    private static void MeshesChanged(IEnumerable<UnityEngine.ProBuilder.ProBuilderMesh> meshes)
+    {
+        if (meshes != null)
+        {
+            foreach (var mesh in meshes)
+            {
+                var version = (ushort)versionInfo.GetValue(mesh);
+
+                if (!versionIndices.TryGetValue(mesh, out var storedVersion))
+                {
+                    versionIndices.Add(mesh, version);
+                }
+                else if (version != storedVersion)
+                {
+                    versionIndices[mesh] = version;
+
+                    var server = mesh.GetComponentInParent<MeshSyncServer>();
+
+                    server?.MeshChanged(mesh);
+                }
+            }
+        }
+    }
+
+    private static void ProBuilderEditor_afterMeshModification(IEnumerable<UnityEngine.ProBuilder.ProBuilderMesh> meshes)
+    {      
+        MeshesChanged(meshes);
+    }
+
+    private static void ProBuilderEditor_selectionUpdated(IEnumerable<UnityEngine.ProBuilder.ProBuilderMesh> meshes)
+    {
+        if (!UnityEditorInternal.InternalEditorUtility.isApplicationActive)
+        {
+            return;
+        }
+
+        MeshesChanged(meshes);
+    }
+
+    private static void ProBuilderBeforeRebuild()
+    {
+        // Change select mode back and forth to ensure probuilder invalidates its internal cache:
+        var t = UnityEditor.ProBuilder.ProBuilderEditor.selectMode;
+        UnityEditor.ProBuilder.ProBuilderEditor.selectMode = UnityEngine.ProBuilder.SelectMode.Object;
+        UnityEditor.ProBuilder.ProBuilderEditor.selectMode = t;
```

Comment:
Yes but then the mode would change for the user when they don't want to. I just need to change the mode because ProBuilder doesn't support an API to refresh the state when the underlying data changed.